include/pybind11/detail/type_caster_base.h

```diff
@@ -827,6 +827,33 @@ using movable_cast_op_type
 template <typename T, typename SFINAE = void>
 struct is_copy_constructible : std::is_copy_constructible<T> {};
 
+// True if Container has a dependent type mapped_type that is equivalent
+// to Container itself
+// Actual implementation in the SFINAE specializations below
+template <typename Container, typename MappedType = Container>
+struct sfinae_is_container_with_self_referential_mapped_type : std::false_type {};
+
+// Tie-breaking between the mapped_type and the value_type specializations is trivial:
+// The specializations are only valid if both conditions are fulfilled:
+// 1) The mapped_type (respectively value_type) exists
+// 2) And it is equivalent to Container
+// So, in each case, only one specialization will activate.
```

Comment:
That's a really neat trick.

What do you think about `is_recursive_container_common` as a name (see other comment for the `is_recursive_container` part).

Reply:
As you noticed below, there is a catch to the trick, so I've split this in two now for structs like
```cpp
struct Self
{
    using mapped_type = Self;
    using value_type = Self;
};
```

I like `is_recursive_container`, using that now